Roles/Crewmate/Enigma.cs

```diff
@@ -0,0 +1,545 @@
+                .SetValueFormat(OptionFormat.Percent);
+            EnigmaGetCluesWithoutReporting = BooleanOptionItem.Create(Id + 16, "EnigmaClueGetCluesWithoutReporting", true, TabGroup.CrewmateRoles, false).SetParent(CustomRoleSpawnChances[CustomRoles.Enigma]);
+
+            OverrideTasksData.Create(Id + 20, TabGroup.CrewmateRoles, CustomRoles.Enigma);
+        }
+        public static void Init()
+        {
+            playerIdList = new();
+            IsEnable = false;
+            ShownClues = new();
+            MsgToSend = new();
+            MsgToSendTitle = new();
+        }
+        public static void Add(byte playerId)
+        {
+            playerIdList.Add(playerId);
+            IsEnable = true;
+            ShownClues.Add(playerId, new List<EnigmaClue>());
+        }
+
+        public static void OnReportDeadBody(PlayerControl player, GameData.PlayerInfo targetInfo)
+        {
+            if (targetInfo == null) return;
+
+            var target = Utils.GetPlayerById(targetInfo.PlayerId);
+            if (target == null) return;
+
+            PlayerControl killer = target.GetRealKiller();
+            if (killer == null) return;
+
+            string title;
+            string msg;
+            var rd = IRandom.Instance;
+
+            foreach (var playerId in playerIdList)
+            {
+                if (!EnigmaGetCluesWithoutReporting.GetBool() && playerId != player.PlayerId) continue;
+
+                var enigmaPlayer = Utils.GetPlayerById(playerId);
+                if (enigmaPlayer == null) continue;
+
+                int tasksCompleted = enigmaPlayer.GetPlayerTaskState().CompletedTasksCount;
+                int stage = 0;
+                bool showStageClue = false;
+
+                if (tasksCompleted >= EnigmaClueStage3Tasks.GetInt())
+                {
+                    stage = 3;
+                    showStageClue = rd.Next(0, 100) < EnigmaClueStage3Probability.GetInt();
+                }
+                else if (tasksCompleted >= EnigmaClueStage2Tasks.GetInt())
+                {
+                    stage = 2;
+                    showStageClue = rd.Next(0, 100) < EnigmaClueStage2Probability.GetInt();
+                }
+                else if (tasksCompleted >= EnigmaClueStage1Tasks.GetInt())
+                    stage = 1;
+
+                var clues = EnigmaClues.Where(a => a.ClueStage <= stage &&
+                    !ShownClues[playerId].Any(b => b.EnigmaClueType == a.EnigmaClueType && b.ClueStage == a.ClueStage))
+                    .ToList();
+                if (clues.Count == 0) continue;
+                if (showStageClue && clues.Any(a => a.ClueStage == stage))
+                    clues = clues.Where(a => a.ClueStage == stage).ToList();
+
+                EnigmaClue clue = clues[rd.Next(0, clues.Count - 1)];
```

Comment:
I'm not sure but `rd.Next()` can't select the last digit

![image](https://github.com/0xDrMoe/TownofHost-Enhanced/assets/104814436/1ec0b60d-2f00-48eb-ae83-e441a21123f1)

Reply:
Thank you Tommy, I didn't know that. I have adjusted it.